Check for contiguous well-formed mask

mikekgfb opened the pull request:
Summary: Check for contiguous well-formed mask

Test Plan: sandcastle, github CI

Reviewed By: frank-wei

Differential Revision: D37301243

jbschlosser commented on torch/nn/modules/transformer.py:
```diff
@@ -203,12 +203,16 @@ def forward(self, src: Tensor, mask: Optional[Tensor] = None, src_key_padding_ma
-                    src.dim() == 3 and self.enable_nested_tensor) :
+                    src.dim() == 3 and self.enable_nested_tensor and
+                    src_key_padding_mask is not None and
+                    # is src_key_padding right polarity or needs inversion?
```
You have it correct here - `src_key_padding` needs inversion.

The mask passed to `_nested_tensor_from_mask_left_aligned` is semantically defined as what to *include* within the NT, and `src_key_padding` specifies locations of padding, which should be *excluded*.

jbschlosser reviewed (commented):
I think we should also have a test that verifies the change. It should satisfy all fast path conditions except the new one (i.e. `src_key_padding` should be a non-left-aligned mask) and verify that the fast path is not taken.

jbschlosser reviewed (approved):
LGTM! Won't block this PR on this but as discussed, we should ensure we have a test for the contiguous, left-aligned mask case as well

Commit:
Check for contiguous well-formed mask (#79927)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/79927

Check for contiguous well-formed mask

Test Plan: sandcastle, github CI

Reviewed By: frank-wei

Differential Revision: D37301243

fbshipit-source-id: 5280991ae21e75f334ce8f4008f6ed1e7f0e1e49
